content/things/js-to-asc.md

```diff
@@ -0,0 +1,563 @@
+
+In my perception, a lot of people think of WebAssembly purely as a performance primitive. It’s compiled, so it’s gotta be fast, right? Well, for the longest time [I have been vocal that WebAssembly and JavaScript have the same _peak_ performance][io19 talk], and I still stand behind that. Given ideal conditions, they both compile to machine code and end up being equally fast. But there’s obviously more nuance here, and when have conditions _ever_ been ideal on the web‽ Instead, I think it would be better if we thought about WebAssembly as a way to get _predictable_ performance.
+
+However, it’s also important to realize that WebAssembly has recently been getting access to performance primitives (like SIMD or shared-memory threads) that JavaScript cannot utilize, giving WebAssembly a _potential_ edge to predictably out-perform JavaScript. There are also some other qualities of WebAssembly that might make it better suited in specific situations than JavaScript:
+
+### No warmup
+
+For JavaScript to be turned into machine code, it needs to run a bit first. In V8’s case, its interpreter “Ignition” is optimized to make code run as _soon_ as possible. Meanwhile, “Sparkplug” compiles your JavaScript to byte code, taking longer but also yieling better performance. Once Sparkplug is done the compiled byte code takes over. While your code is running (either through Ignition or Sparkplug), it is closely observed by V8. It takes not of what kind of type of data you store in your variables, function parameters and so on. Once sufficient data has been collected, V8’s optimizing compiler “TurboFan” kicks in and generates low-level machine code using that type data. This will give a significant speed boost.
+
+WebAssembly, on the other hand, is strongly typed. It can be turned into machine code _straight away_. V8 has a streaming Wasm compiler called “Liftoff“ which, like Ignition, is optimized to generate machine code _fast_, making sure your WebAssembly can start working as soon as possible, at the cost of generating potentially suboptimal machine code. The second Liftoff is done, TurboFan kicks in and both generates optimized machine code that will run faster than what Liftoff produced, but will take  longer to generate. The big difference to JavaScript is that the TurboFan can do its work without having to observe your Wasm first.
+
+### No tierdown
+
+The machine code that TurboFan generates for JavaScript is only usable for as long as the assumptions about types hold. If TurboFan generated machine code for a funtion `f` with a number as a parameter, and now all of the sudden that function `f` gets called with an object, the engine has to fall back to Ignition or Sparkplug. That’s called a “deoptimization” (or “deopt” for short). Again, because WebAssembly is strongly typed, the types _can’t_ change. Not only that, but they types that WebAssembly supports are designed to map well to machine code. Deopts can’t happen with WebAssembly.
```

Comment:
Typo: the**y** types that WebAssembly supports

Reply:
Cheers